Update contribution guidelines (#1029)
genomic-medicine-sweden/nallo: .github/CONTRIBUTING.md

#### Pull Requests

When opening a pull request to suggest changes to the code, please make sure to follow the [Pipeline contribution conventions](#pipeline-contribution-conventions) for the code and to fill in the necessary information in the pull request template as well as address all points in the `PR checklist`.

##### Review

When reviewing a PR, make sure to check that:

- The code follows the [Pipeline contribution conventions](#pipeline-contribution-conventions).
- The information in the PR (and related issue) is clear and sufficient to understand the change and the motivation for it - title, description and entry in `CHANGELOG.md`, if applicable.
- All the items in the `PR checklist` have been addressed, the changes are well documented and the tests are passing.

Be positive and constructive in your review, and whenever possible offer suggestions for improvement rather than just pointing out issues.

### Software versioning, changelog and updates

#### Semantic versioning and changelog

> Release versioning tries to be maintained according to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and a changelog is maintained according to the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format. Altough, for a Nextflow pipeline it's hard do decide what is a breaking/non-breaking API change.

Reviewer: ```suggestion
Release versioning tries to be maintained according to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and a changelog is maintained according to the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format. Although, for a Nextflow pipeline it can be hard to decide what is a breaking/non-breaking API change.
```

Author: Thanks! I didn't particularly like this sentence before, but I also don't think Semantic Versioning really fits a Nextflow pipeline...hard (for me) to come up with good guidelines for it.

> The `Fixed` section of the changelog should be reserved for bugs fixed from one release to another, but bug fixes that appeared and was solved in dev. These should go in the `Changed` section.

Reviewer 2: I am not sure i understand this sentence, do you mean ", but *not* for bug fixes that appeared and were solved in dev" ?

> The `Fixed` section of the changelog should be reserved for bugs fixed from one release to another, but bug fixes that appeared and was solved in dev. These should go in the `Changed` section.

Reviewer: ```suggestion
Bugs fixed between releases belong in the `Fixed`  section of the changelog, whereas bug fixes that appeared and were solved in dev should go in the `Changed` section.
```

##### Patch

:warning: Only in the unlikely and regretful event of a release happening with a bug.

- On the genomic-medicine-sweden/nallo repository, make a new branch `patch` based on `upstream/main` or `upstream/master`.
- Fix the bug, and bump version (X.Y.Z+1).
- Open a pull-request from `patch` to `main`/`master` with the changes.

#### Nextflow version bumping

You may bump the minimum required version of nextflow in the pipeline with: `nf-core pipelines bump-version --nextflow . [min-nf-version]`

#### Update nf-core template

Since this is not an nf-core pipeline, the nf-core template is not automatically updated in the `TEMPLATE` branch. Follow these step to update the template:

1. Update the `TEMPLATE` branch by running `nf-core pipelines sync`. Fix any merge conflicts and open a PR to then merge the changes.

> 1. Open a PR to merge the `TEMPLATE` branch into `dev` to update the template files in the main codebase.

Reviewer: ```suggestion
1. Update the `TEMPLATE` branch by running `nf-core pipelines sync`. Fix any merge conflicts.
2. Open a PR to merge the `TEMPLATE` branch into `dev` to update the template files in the main codebase.
```

### Developer setup

#### Installation and dependencies for development

In order to run the pipeline, develop and test your changes locally, we recommend that you set up:

- A conda environment with `nextflow`, `nf-core` tools and `nf-test`. - For this, follow the instructions from the [nf-core documentation](https://nf-co.re/docs/nf-core-tools/cli/installation#install-with-conda) and install `nf-test` from bioconda or by following the [nf-test installation instructions](https://www.nf-test.com/installation/) . Additional information about [installation of nf-core dependencies](https://nf-co.re/docs/usage/getting_started/installation/) is also available, if needed.
- Install Docker (https://www.docker.com/products/docker-desktop/) and make sure the daemon is running when you want to run the tests locally.

Make sure to [test your setup](https://nf-co.re/docs/usage/introduction#how-to-run-a-pipeline) with `-profile test` before running the workflow on actual data.

> Additionally, pre-commit hooks are set up to automatically check the code and generate parameters documentation when committing. To install the pre-commit hooks, run `pre-commit install` in the root of the repository. Note that, other than the default pre-commit hooks from the nf-core template, there is an additional hook - [`nf-core pipelines schema docs` pre-commit hook](https://github.com/genomic-medicine-sweden/nf-core-schema-docs) - set up in this codebase for automatically generating parameters documentation if there are any changes to parameters in your commit.

Reviewer: ```suggestion
Additionally, pre-commit hooks should be set up to automatically check the code and generate parameters documentation when committing. To install the pre-commit hooks, run `pre-commit install` in the root of the repository. Besides the default nf-core hooks, this repository includes the [nf-core pipelines schema docs](https://github.com/genomic-medicine-sweden/nf-core-schema-docs) hook. It automatically updates parameter documentation whenever your commit includes changes to `nextflow_schema.json`.
```

- Every subworkflow should have real tests and `-stub` tests at `subworkflows/local/<name>/tests/main.nf.test`.
- Snapshot files (`*.nf.test.snap`) are committed alongside tests — update them when outputs change.
- Pipeline-level tests live in `tests/`.
- When modules in tests require different parameters, use `params { <module>_args = ... }` similar to https://nf-co.re/docs/specifications/components/modules/testing#configuration-of-extargs-in-tests instead of having mutliple configs.

### Style

- Both `take:` and `emit:` block entries require an inline type comment. Use `name // type: [mandatory|optional] description` for `take:` and `name = value // channel: [type description]` for `emit:`. Always include the comment — never leave an entry uncommented.

  ```groovy
  take:
  ch_vcf                // channel: [mandatory] [ val(meta), path(vcf) ]
  ch_reduced_penetrance // channel: [optional]  [ path(penetrance) ]
  val_aligner           // string:  [mandatory] aligner name (bwa/bwamem2/bwameme)
  process_with_sort     // Boolean

  emit:
  vcf     = ch_vcf      // channel: [ val(meta), path(vcf) ]
  publish = ch_publish  // channel: [ val(destination), val(value) ]
  ```

> - Avoid using `ch_* = <...>`, use the `.set {ch_*}` operator to create new channels whenever possible.

Author: Open to change this, I think 😄
